qiime2/metadata/metadata.py

```diff
@@ -87,6 +89,20 @@ def artifacts(self):
         """
         return tuple(self._artifacts)
 
+    @property
+    def non_artifact_metadata(self):
+        """Non artifacts that are the source of the metadata.
```

Comment:
Maybe we don't need this property at all? One option is to refactor this and the `artifacts` property above as a new `source` (or similar) property. That property could expose some new list-like data made up of records that keep track of the data, and the type of data (native object vs Artifact). I think this might allow you to clean up some logic down below (but, I haven't looked too closely). The pre-existing `artifacts` property could remain, but it could just return a filtered view of the `source` property.

Again, just spitballing an idea, but, might be worth looking into.

Reply:
I've been playing with that idea, and I think I like it

Reply:
Okay, how about we keep track of three separate sources:

- `artifact_sources`
- `dataframe_sources`
- `tsv_sources`

?

I haven't looked too closely at the constructors in a while, so I might be forgetting some "gotcha" w.r.t disambiguating the dataframe and tsv sources, but my hunch is that with some updates to the constructors, we can keep track of "who is who". I think this will be useful, especially down the line.

Reply:
Would we have three separate lists for those three things or would we still have only a single `source` list that we would filter those out of?

Reply:
It shouldn't be to hard to differentiate between those three sources either by the way. Sources from dataframes call the `Metadata` constructor and pass it a `DataFrame` while sources from tsvs call `Metadata.load` on a tsv. Checking if it came from an artifact can be done just by typechecking.

Reply:
Previously we were adding the `Metadata` objects to the `sources` list for all `non_artifact` sources. If we're going to distinguish between .tsvs and `DataFrame` object sources now, should we be adding something that more accurately reflects those to the `sources` list. Like say the actual `DataFrame` object for `Dataframe` sources and the filepath to the .tsv for .tsv sources or something?

Reply:
@thermokarst https://github.com/qiime2/qiime2/pull/464/commits/8ef8dda515213b43cbc4aa1cde5b7e9f5c13f528 this commit begins to roughly implement what we discussed above. Not really complete or tested, but something along those lines.

Reply:
Thanks @Oddant1, I will respond inline.